google/cloud/bigtable/batcher.py

```diff
@@ -13,104 +13,252 @@
+        self.exc = exc
+        self.message = message
+        super().__init__(self.message)
+
+
+class _MutationsBatchQueue(object):
+    """Private Threadsafe Queue to hold rows for batching."""
+
+    def __init__(self, max_mutation_bytes=MAX_MUTATION_SIZE, flush_count=FLUSH_COUNT):
+        """Specify the queue constraints"""
+        self._queue = queue.Queue()
+        self.total_mutation_count = 0
+        self.total_size = 0
+        self.max_mutation_bytes = max_mutation_bytes
+        self.flush_count = flush_count
+
+    def get(self):
+        """Retrieve an item from the queue. Recalculate queue size."""
+        row = self._queue.get()
+        mutation_size = row.get_mutations_size()
+        self.total_mutation_count -= len(row._get_mutations())
+        self.total_size -= mutation_size
+        return row
+
+    def put(self, item):
+        """Insert an item to the queue. Recalculate queue size."""
+
+        mutation_count = len(item._get_mutations())
+
+        self._queue.put(item)
+
+        self.total_size += item.get_mutations_size()
+        self.total_mutation_count += mutation_count
+
+    def full(self):
+        """Check if the queue is full."""
+        if (
+            self.total_mutation_count >= self.flush_count
+            or self.total_size >= self.max_mutation_bytes
+        ):
+            return True
+        return False
+
+    def empty(self):
+        return self._queue.empty()
+
+
+@dataclass
+class _BatchInfo:
+    """Keeping track of size of a batch"""
+
+    mutations_count: int = 0
+    rows_count: int = 0
+    mutations_size: int = 0
+
+
+class FlowControl(object):
+    def __init__(
+        self,
+        max_mutations=MAX_OUTSTANDING_ELEMENTS,
+        max_mutation_bytes=MAX_OUTSTANDING_BYTES,
+    ):
+        """Control the inflight requests. Keep track of the mutations, row bytes and row counts.
+        As requests to backend are being made, adjust the number of mutations being processed.
+
+        If threshold is reached, block the flow.
+        Reopen the flow as requests are finished.
+        """
+        self.max_mutations = max_mutations
+        self.max_mutation_bytes = max_mutation_bytes
+        self.inflight_mutations = 0
+        self.inflight_size = 0
+        self.event = threading.Event()
+        self.event.set()
+
+    def is_blocked(self):
+        """Returns True if:
+
+        - inflight mutations >= max_mutations, or
+        - inflight bytes size >= max_mutation_bytes, or
+        """
+
+        return (
+            self.inflight_mutations >= self.max_mutations
+            or self.inflight_size >= self.max_mutation_bytes
+        )
+
+    def control_flow(self, batch_info):
+        """
+        Calculate the resources used by this batch
+        """
+
+        self.inflight_mutations += batch_info.mutations_count
+        self.inflight_size += batch_info.mutations_size
+        self.set_flow_control_status()
+        self.wait()
```

Comment:
I adjusted the logic, moving the `wait` to the `flush_async` function. If the batch is causing the event to be blocked, then it will be sent through, but the subsequent thread will be blocked and waited.